File: 贪心算法/middle/134.加油站.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <unordered_set>
#include <unordered_map>
#include <math.h>

using namespace std;

/*
 * @lc app=leetcode.cn id=134 lang=cpp
 *
 * [134] 加油站
 */

// @lc code=start
class Solution
{
public:
    int canCompleteCircuit(vector<int> &gas, vector<int> &cost)
    {
        int curSum = 0;
        int min = INT_MAX; // 从起点出发，油箱里的油量最小值

        for (int i = 0; i < cost.size(); i++)
        {
            curSum += gas[i] - cost[i];

            if (min > curSum)
            {
                min = curSum;
            }
        }

        // 总消耗大于总油量，则跑不了一圈
        if (curSum < 0)
        {
            return -1;
        }

        // 从0开始跑一直是正数，则每个阶段都满足，直接从0开始跑就行
        if (min >= 0)
        {
            return 0;
        }

        for (int i = cost.size() - 1; i > 0; i--)
        {
            int rest = gas[i] - cost[i];
            min += rest;

            // 从后往前遍历，如果累积的油量能把这个最大的负数填平
            if (min >= 0)
            {
                // 则从i位置开始跑就行
                return i;
            }
        }

        return -1;
    }
};
// @lc code=end

File: 贪心算法/middle/134.加油站.ts
/*
 * @lc app=leetcode.cn id=134 lang=typescript
 *
 * [134] 加油站
 */

// @lc code=start
function canCompleteCircuit(gas: number[], cost: number[]): number {
  let currentSum = 0
  let totalSum = 0
  let startIndex = 0

  for (let i = 0; i < gas.length; i++) {
    // 到当前加油站为止剩余的油量
    currentSum += gas[i] - cost[i]
    // 记录跑完所有加油站的剩余油量
    totalSum += gas[i] - cost[i]

    // 如果之前的剩余油量为负数，说明之前0到i个加油站作为起始位置，都跑不到下一个加油站
    // 例如currentSum[0] <0 ,index=0的加油站不行
    // currentSum[0] > 0 currentSum[1] < 0, ,index=1的加油站不行，以此类推
    if (currentSum < 0) {
      // 新的起始加油站下标为i+1
      startIndex = i + 1
      // 之前的和清零
      currentSum = 0
    }
  }

  // 总和都小于0，那么一不完一圈
  if (totalSum < 0) {
    return -1
  }

  return startIndex
};
// @lc code=end

